src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/TestRequestHandler.cs

```diff
@@ -19,306 +19,375 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
-            this.transport = new SocketTransport(communicationManager, connectionInfo);
+            this.communicationEndPoint = communicationEndpoint;
+            this.connectionInfo = connectionInfo;
             this.dataSerializer = dataSerializer;
+            this.requestSenderConnected = new ManualResetEventSlim(false);
+            this.testHostManagerFactoryReady = new ManualResetEventSlim(false);
+            this.sessionCompleted = new ManualResetEventSlim(false);
+            this.onAckMessageRecieved = onAckMessageRecieved;
+            this.jobQueue = jobQueue;
         }
 
-        /// <inheritdoc/>
-        public void InitializeCommunication()
+        protected TestRequestHandler(TestHostConnectionInfo connectionInfo, IDataSerializer dataSerializer)
         {
-            this.transport.Initialize();
-        }
-
-        /// <inheritdoc/>
-        public bool WaitForRequestSenderConnection(int connectionTimeout)
-        {
-            return this.transport.WaitForConnection(connectionTimeout);
-        }
+            this.connectionInfo = connectionInfo;
+            this.dataSerializer = dataSerializer;
+            this.requestSenderConnected = new ManualResetEventSlim(false);
+            this.sessionCompleted = new ManualResetEventSlim(false);
+            this.testHostManagerFactoryReady = new ManualResetEventSlim(false);
+            this.onAckMessageRecieved = (message) => { throw new NotImplementedException(); };
 
-        /// <summary>
-        /// Listens to the commands from server
-        /// </summary>
-        /// <param name="testHostManagerFactory">the test host manager.</param>
-        public void ProcessRequests(ITestHostManagerFactory testHostManagerFactory)
-        {
-            bool isSessionEnd = false;
+            this.SetCommunicationEndPoint();
 
-            var jobQueue = new JobQueue<Action>(
-                action => { action(); },
+            this.jobQueue = new JobQueue<Action>(
+                (action) => { action(); },
                 "TestHostOperationQueue",
                 500,
                 25000000,
                 true,
                 (message) => EqtTrace.Error(message));
+        }
 
-            do
+        /// <inheritdoc />
+        public virtual void InitializeCommunication()
+        {
+            this.communicationEndPoint.Connected += (sender, connectedArgs) =>
             {
-                var message = this.communicationManager.ReceiveMessage();
-                if (EqtTrace.IsInfoEnabled)
+                if (!connectedArgs.Connected)
                 {
-                    EqtTrace.Info("TestRequestHandler.ProcessRequests: received message: {0}", message);
+                    requestSenderConnected.Set();
+                    throw connectedArgs.Fault;
                 }
+                this.channel = connectedArgs.Channel;
+                this.channel.MessageReceived += this.OnMessageReceived;
```

Comment:
>MessageReceived [](start = 29, length = 15)

can the processing of a second message begin before the completion of an earlier one..? what is the impact..?